### PR TITLE
FM2-481: add ObservationSearchParams class and refactor Observation service search method

### PR DESCRIPTION
**Pull request by pmanko2**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.HashSet;
+
+import ca.uhn.fhir.model.api.Include;
+import ca.uhn.fhir.rest.api.SortSpec;
+import ca.uhn.fhir.rest.param.DateRangeParam;
+import ca.uhn.fhir.rest.param.TokenAndListParam;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java:22]: not thrilled about SuperBuilder still being part of the experimental package. may look into refactoring to no longer use SuperBuilder if possible. can probably use the alternative method described here https://www.baeldung.com/lombok-builder-inheritance to still make use of inheritance for resource search params

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Class containing shared search (and search result) parameters that apply to all resources See
+ * https://www.hl7.org/fhir/search.html#Summary for a comprehensive description of these parameters
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@SuperBuilder
+public class BaseResourceSearchParams implements Serializable {
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java:32]: did a first pass at pulling out params that are used in most resources, although it does seem like resources explicitly define which params they accept within the resource Provider class directly. wondering if this means that not all resources will support the common parameters defined in https://www.hl7.org/fhir/search.html#Summary ?

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/ObservationSearchParams.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	
+	private TokenAndListParam code;
+	
+	private TokenAndListParam category;
+	
+	/**
+	 * Custom AllArgsConstructor (instead of using @AllArgsConstructor annotation) that supports super
+	 * class instantiation
+	 */
+	public ObservationSearchParams(ReferenceAndListParam encounterReference, ReferenceAndListParam patientReference,
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/param/ObservationSearchParams.java:50]: lombok doesn't support super class field instantiation via the AllArgsConstructor annotation (https://github.com/projectlombok/lombok/issues/1229), so if we want to go this route we'll need to define our own, for each class that inherits from BaseResourceSearchParams

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java (hunk 2 above)
Comment (pmanko2) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java:32]: also technically we have common search params and common search result params, not sure if it's better to distinguish between those two and have BaseResourceSearchParams/BaseResourceSearchResultParams

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/SearchParameterMap.java [hunk 4]
```diff
@@ -56,10 +56,6 @@ public SearchParameterMap addParameter(@Nonnull String key, @Nonnull Serializabl
 	 * @return {@link org.openmrs.module.fhir2.api.search.param.SearchParameterMap}
 	 */
 	public SearchParameterMap addParameter(@Nonnull String key, String propertyName, @Nonnull Serializable param) {
-		if (key == null || param == null) {
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/param/SearchParameterMap.java:59]: ide may have accidentally changed this, but it does seem like it's an unnecessary if statement with the Nonnull annotations (unless those aren't guaranteed to throw an exception in some cases?)

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderWebTest.java [hunk 5]
```diff
@@ -38,6 +38,7 @@
 import java.util.Objects;
 
 import ca.uhn.fhir.model.api.Include;
+import ca.uhn.fhir.rest.api.SortSpec;
```
Comment (pmanko2) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderWebTest.java:41]: these test changes were not fun

Comment (pmanko2) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderWebTest.java:41]: also seems like a lot of the r3/r4 tests are duplicated, wondering if it might be worth pulling out the duplicates into a base `ObservationFhirResourceProviderWebTest` test class and only include r3 vs r4 specific testing in these classes?

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java (hunk 1 above)
Comment (pmanko2) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java]: ^ ended up going this route

File: api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQueryInclude.java [hunk 6]
```diff
@@ -216,13 +217,14 @@ private IBundleProvider handleLocationReverseInclude(ReferenceAndListParam param
...
 	}
 	
-	private IBundleProvider handleObservationReverseInclude(ReferenceAndListParam params, String targetType) {
+	private IBundleProvider handleObservationReverseInclude(ReferenceAndListParam param, String targetType) {
 		switch (targetType) {
 			case FhirConstants.OBSERVATION:
-				return observationService.searchForObservations(null, null, params, null, null, null, null, null, null, null,
-				    null, null, null, null, null);
+				ObservationSearchParams observationSearchParams = new ObservationSearchParams();
+				observationSearchParams.setHasMember(param);
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQueryInclude.java:224]: verify this is the right search param to set

**Review by pmanko — APPROVED** [review]
Looks good, thanks @pmanko2! 

The logic for processing and mapping search parameters still is quite complex, since we seem to be doing a subsequent mappings between different representations of the parameters for each Resource. 

This is my first time taking a close look at the search functionality, so perhaps it would be useful to bring this up and discuss in the next FHIR meeting a larger group and get some perspective from others.

**Comment by pmanko2** [comment]
> Looks good, thanks @pmanko2!
> 
> The logic for processing and mapping search parameters still is quite complex, since we seem to be doing a subsequent mappings between different representations of the parameters for each Resource.
> 
> This is my first time taking a close look at the search functionality, so perhaps it would be useful to bring this up and discuss in the next FHIR meeting a larger group and get some perspective from others.

Yeah agreed, i'd love to discuss whether we need both "SearchParameter" classes and the SearchParameterMap class that is used within the daos to create criteria queries. This implementation allows us to define a single place that a SearchParameterMap object is configured for each Resource. We don't have to re-specify the different key/property constants required by SearchParamMap anytime we want to make a Resource search request. However, there are already multiple services that accept a SearchParameterMap as the only parameter for their search method, so it might be worth exploring standardizing around that.

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java (hunk 2 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java]: Hmmmm... yeah. It might make sense to implement some common parameters, but I'd want to discuss how we do it. The only two common search parameters that are unproblematically acceptable for all types are `_id` and `_lastUpdated`.

File: api/src/main/java/org/openmrs/module/fhir2/api/FhirObservationService.java [hunk 7]
```diff
@@ -9,35 +9,21 @@
...
-	        SortSpec sort, HashSet<Include> includes, HashSet<Include> revIncludes);
+	IBundleProvider searchForObservations(ObservationSearchParams observationSearchParams);
 	
-	IBundleProvider getLastnObservations(NumberParam max, ReferenceAndListParam patientReference, TokenAndListParam category,
-	        TokenAndListParam code);
+	IBundleProvider getLastnObservations(NumberParam max, ObservationSearchParams observationSearchParams);
 	
-	IBundleProvider getLastnEncountersObservations(NumberParam max, ReferenceAndListParam patientReference,
-	        TokenAndListParam category, TokenAndListParam code);
+	IBundleProvider getLastnEncountersObservations(NumberParam max, ObservationSearchParams observationSearchParams);
```
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/FhirObservationService.java:28]: That's fine. For `$lastn` I think we only implemented [the minimum requirements](https://www.hl7.org/fhir/observation-operation-lastn.html).

**Comment by ibacher** [comment]
> i'd love to discuss whether we need both "SearchParameter" classes and the SearchParameterMap class

We don't, but I thought I'd make this a series of smaller steps, e.g., we add the `SearchParameter` classes and once those are all in place we can kibosh `SearchParameterMap` altogether. It has the bonus of having real types.

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java (hunk 2 above)
Comment (pmanko2) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java]: are you ok with me leaving the common params as is for now? sounds like we'll at least have _id and _lastUpdated in that class, so imo it can stick around even if we pull out some params that are currently defined there in the future.

**Comment by pmanko2** [comment]
@ibacher fixed up the import issues, let me know if there's anything else that needs to be looked at
